File: sdk/typescript/src/contacts/correspondence.ts
export type CorrespondenceChannel = "email" | "sms" | "imessage" | "calls";
export type CorrespondenceContentMode = "metadata" | "preview" | "full";
export type CorrespondenceTranscriptMode = "none" | "abridged" | "full";
export type CorrespondenceOrder = "asc" | "desc";
export type CorrespondenceChannelStatus = "available" | "no_identifier" | "no_resource";
export type CorrespondenceDirection = "inbound" | "outbound";

export interface CorrespondenceMediaMetadata {
  count: number;
}

export interface CorrespondenceAttachmentMetadata {
  filename: string | null;
  contentType: string | null;
  size: number | null;
}

export interface CorrespondenceTranscriptEntry {
  id: string | null;
  seq: number | null;
  party: string | null;
  text: string | null;
  tsMs: number | null;
  marker: "abridged" | null;
  omittedTurns: number | null;
  omittedMs: number | null;
}

export interface CorrespondenceItemBase {
  sourceId: string;
  direction: CorrespondenceDirection;
  occurredAt: Date;
  identityId: string;
  status: string | null;
  detailUrl: string | null;
}

export interface EmailCorrespondenceItem extends CorrespondenceItemBase {
  channel: "email";
  mailboxEmail: string;
  threadId: string | null;
  fromAddress: string;
  toAddresses: string[];
  ccAddresses: string[];
  bccAddresses: string[];
  subject: string | null;
  snippet: string | null;
  bodyText: string | null;
  contentUnavailable: boolean;
  attachments: CorrespondenceAttachmentMetadata[];
}

export interface SmsCorrespondenceItem extends CorrespondenceItemBase {
  channel: "sms";
  conversationId: string;
  localResourceId: string;
  localPhoneNumber: string;
  senderPhoneNumber: string | null;
  participants: string[];
  matchedContactPhone: string;
  isGroup: boolean;
  text: string | null;
  media: CorrespondenceMediaMetadata | null;
}

export interface IMessageCorrespondenceItem extends CorrespondenceItemBase {
  channel: "imessage";
  conversationId: string;
  remoteHandle: string;
  service: string;
  text: string | null;
  media: CorrespondenceMediaMetadata | null;
}

export interface CallCorrespondenceItem extends CorrespondenceItemBase {
  channel: "calls";
  remotePhoneNumber: string;
  localPhoneNumber: string | null;
  startedAt: Date | null;
  endedAt: Date | null;
  durationSeconds: number | null;
  transcript: CorrespondenceTranscriptEntry[] | null;
  transcriptAbridged: boolean;
  transcriptUnavailable: boolean;
}

export type CorrespondenceItem =
  | EmailCorrespondenceItem
  | SmsCorrespondenceItem
  | IMessageCorrespondenceItem
  | CallCorrespondenceItem;

export interface CorrespondenceChannelResult {
  channel: CorrespondenceChannel;
  status: CorrespondenceChannelStatus;
  returned: number;
}

export interface ContactCorrespondence {
  contactId: string;
  identityId: string;
  items: CorrespondenceItem[];
  channels: CorrespondenceChannelResult[];
  nextCursor: string | null;
}

interface RawCorrespondenceItemBase {
  source_id: string;
  direction: CorrespondenceDirection;
  occurred_at: string;
  identity_id: string;
  status: string | null;
  detail_url: string | null;
}

export interface RawEmailCorrespondenceItem extends RawCorrespondenceItemBase {
  channel: "email";
  mailbox_email: string;
  thread_id: string | null;
  from_address: string;
  to_addresses: string[];
  cc_addresses: string[];
  bcc_addresses: string[];
  subject: string | null;
  snippet: string | null;
  body_text: string | null;
  content_unavailable: boolean;
  attachments: Array<{
    filename: string | null;
    content_type: string | null;
    size: number | null;
  }>;
}

export interface RawSmsCorrespondenceItem extends RawCorrespondenceItemBase {
  channel: "sms";
  conversation_id: string;
  local_resource_id: string;
  local_phone_number: string;
  sender_phone_number: string | null;
  participants: string[];
  matched_contact_phone: string;
  is_group: boolean;
  text: string | null;
  media: CorrespondenceMediaMetadata | null;
}

export interface RawIMessageCorrespondenceItem extends RawCorrespondenceItemBase {
  channel: "imessage";
  conversation_id: string;
  remote_handle: string;
  service: string;
  text: string | null;
  media: CorrespondenceMediaMetadata | null;
}

export interface RawCallCorrespondenceItem extends RawCorrespondenceItemBase {
  channel: "calls";
  remote_phone_number: string;
  local_phone_number: string | null;
  started_at: string | null;
  ended_at: string | null;
  duration_seconds: number | null;
  transcript: Array<{
    id: string | null;
    seq: number | null;
    party: string | null;
    text: string | null;
    ts_ms: number | null;
    marker: "abridged" | null;
    omitted_turns: number | null;
    omitted_ms: number | null;
  }> | null;
  transcript_abridged: boolean;
  transcript_unavailable: boolean;
}

export type RawCorrespondenceItem =
  | RawEmailCorrespondenceItem
  | RawSmsCorrespondenceItem
  | RawIMessageCorrespondenceItem
  | RawCallCorrespondenceItem;

export interface RawContactCorrespondence {
  contact_id: string;
  identity_id: string;
  items: RawCorrespondenceItem[];
  channels: Array<{
    channel: CorrespondenceChannel;
    status: CorrespondenceChannelStatus;
    returned: number;
  }>;
  next_cursor: string | null;
}

function parseBase(r: RawCorrespondenceItemBase): CorrespondenceItemBase {
  return {
    sourceId: r.source_id,
    direction: r.direction,
    occurredAt: new Date(r.occurred_at),
    identityId: r.identity_id,
    status: r.status,
    detailUrl: r.detail_url,
  };
}

export function parseCorrespondenceItem(r: RawCorrespondenceItem): CorrespondenceItem {
  const base = parseBase(r);
  switch (r.channel) {
    case "email":
      return {
        ...base,
        channel: r.channel,
        mailboxEmail: r.mailbox_email,
        threadId: r.thread_id,
        fromAddress: r.from_address,
        toAddresses: r.to_addresses,
        ccAddresses: r.cc_addresses,
        bccAddresses: r.bcc_addresses,
        subject: r.subject,
        snippet: r.snippet,
        bodyText: r.body_text,
        contentUnavailable: r.content_unavailable,
        attachments: r.attachments.map((a) => ({
          filename: a.filename,
          contentType: a.content_type,
          size: a.size,
        })),
      };
    case "sms":
      return {
        ...base,
        channel: r.channel,
        conversationId: r.conversation_id,
        localResourceId: r.local_resource_id,
        localPhoneNumber: r.local_phone_number,
        senderPhoneNumber: r.sender_phone_number,
        participants: r.participants,
        matchedContactPhone: r.matched_contact_phone,
        isGroup: r.is_group,
        text: r.text,
        media: r.media,
      };
    case "imessage":
      return {
        ...base,
        channel: r.channel,
        conversationId: r.conversation_id,
        remoteHandle: r.remote_handle,
        service: r.service,
        text: r.text,
        media: r.media,
      };
    case "calls":
      return {
        ...base,
        channel: r.channel,
        remotePhoneNumber: r.remote_phone_number,
        localPhoneNumber: r.local_phone_number,
        startedAt: r.started_at ? new Date(r.started_at) : null,
        endedAt: r.ended_at ? new Date(r.ended_at) : null,
        durationSeconds: r.duration_seconds,
        transcript: r.transcript?.map((t) => ({
          id: t.id,
          seq: t.seq,
          party: t.party,
          text: t.text,
          tsMs: t.ts_ms,
          marker: t.marker,
          omittedTurns: t.omitted_turns,
          omittedMs: t.omitted_ms,
        })) ?? null,
        transcriptAbridged: r.transcript_abridged,
        transcriptUnavailable: r.transcript_unavailable,
      };
  }
}

export function parseContactCorrespondence(r: RawContactCorrespondence): ContactCorrespondence {
  return {
    contactId: r.contact_id,
    identityId: r.identity_id,
    items: r.items.map(parseCorrespondenceItem),
    channels: r.channels,
    nextCursor: r.next_cursor,
  };
}

File: sdk/rust/src/contacts/correspondence.rs
//! Unified contact correspondence models.

use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum CorrespondenceChannel {
    Email,
    Sms,
    IMessage,
    Calls,
}

impl CorrespondenceChannel {
    pub(crate) const fn as_str(self) -> &'static str {
        match self {
            Self::Email => "email",
            Self::Sms => "sms",
            Self::IMessage => "imessage",
            Self::Calls => "calls",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "lowercase")]
pub enum CorrespondenceContentMode {
    Metadata,
    #[default]
    Preview,
    Full,
}

impl CorrespondenceContentMode {
    pub(crate) const fn as_str(self) -> &'static str {
        match self {
            Self::Metadata => "metadata",
            Self::Preview => "preview",
            Self::Full => "full",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "lowercase")]
pub enum CorrespondenceTranscriptMode {
    #[default]
    None,
    Abridged,
    Full,
}

impl CorrespondenceTranscriptMode {
    pub(crate) const fn as_str(self) -> &'static str {
        match self {
            Self::None => "none",
            Self::Abridged => "abridged",
            Self::Full => "full",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "lowercase")]
pub enum CorrespondenceOrder {
    Asc,
    #[default]
    Desc,
}

impl CorrespondenceOrder {
    pub(crate) const fn as_str(self) -> &'static str {
        match self {
            Self::Asc => "asc",
            Self::Desc => "desc",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CorrespondenceChannelStatus {
    Available,
    NoIdentifier,
    NoResource,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum CorrespondenceDirection {
    Inbound,
    Outbound,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CorrespondenceMediaMetadata {
    pub count: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CorrespondenceAttachmentMetadata {
    #[serde(default)]
    pub filename: Option<String>,
    #[serde(default)]
    pub content_type: Option<String>,
    #[serde(default)]
    pub size: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CorrespondenceTranscriptEntry {
    #[serde(default)]
    pub id: Option<Uuid>,
    #[serde(default)]
    pub seq: Option<u64>,
    #[serde(default)]
    pub party: Option<String>,
    #[serde(default)]
    pub text: Option<String>,
    #[serde(default)]
    pub ts_ms: Option<u64>,
    #[serde(default)]
    pub marker: Option<CorrespondenceTranscriptMarker>,
    #[serde(default)]
    pub omitted_turns: Option<u64>,
    #[serde(default)]
    pub omitted_ms: Option<u64>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum CorrespondenceTranscriptMarker {
    Abridged,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CorrespondenceItemBase {
    pub source_id: Uuid,
    pub direction: CorrespondenceDirection,
    pub occurred_at: String,
    pub identity_id: Uuid,
    #[serde(default)]
    pub status: Option<String>,
    #[serde(default)]
    pub detail_url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EmailCorrespondenceItem {
    #[serde(flatten)]
    pub common: CorrespondenceItemBase,
    pub mailbox_email: String,
    #[serde(default)]
    pub thread_id: Option<Uuid>,
    pub from_address: String,
    pub to_addresses: Vec<String>,
    #[serde(default)]
    pub cc_addresses: Vec<String>,
    #[serde(default)]
    pub bcc_addresses: Vec<String>,
    #[serde(default)]
    pub subject: Option<String>,
    #[serde(default)]
    pub snippet: Option<String>,
    #[serde(default)]
    pub body_text: Option<String>,
    #[serde(default)]
    pub content_unavailable: bool,
    #[serde(default)]
    pub attachments: Vec<CorrespondenceAttachmentMetadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SmsCorrespondenceItem {
    #[serde(flatten)]
    pub common: CorrespondenceItemBase,
    pub conversation_id: Uuid,
    pub local_resource_id: Uuid,
    pub local_phone_number: String,
    #[serde(default)]
    pub sender_phone_number: Option<String>,
    pub participants: Vec<String>,
    pub matched_contact_phone: String,
    pub is_group: bool,
    #[serde(default)]
    pub text: Option<String>,
    #[serde(default)]
    pub media: Option<CorrespondenceMediaMetadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct IMessageCorrespondenceItem {
    #[serde(flatten)]
    pub common: CorrespondenceItemBase,
    pub conversation_id: Uuid,
    pub remote_handle: String,
    pub service: String,
    #[serde(default)]
    pub text: Option<String>,
    #[serde(default)]
    pub media: Option<CorrespondenceMediaMetadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CallCorrespondenceItem {
    #[serde(flatten)]
    pub common: CorrespondenceItemBase,
    pub remote_phone_number: String,
    #[serde(default)]
    pub local_phone_number: Option<String>,
    #[serde(default)]
    pub started_at: Option<String>,
    #[serde(default)]
    pub ended_at: Option<String>,
    #[serde(default)]
    pub duration_seconds: Option<u64>,
    #[serde(default)]
    pub transcript: Option<Vec<CorrespondenceTranscriptEntry>>,
    #[serde(default)]
    pub transcript_abridged: bool,
    #[serde(default)]
    pub transcript_unavailable: bool,
}

/// A correspondence item parsed by its `channel` discriminator.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "channel", rename_all = "lowercase")]
pub enum CorrespondenceItem {
    Email(EmailCorrespondenceItem),
    Sms(SmsCorrespondenceItem),
    IMessage(IMessageCorrespondenceItem),
    Calls(CallCorrespondenceItem),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CorrespondenceChannelResult {
    pub channel: CorrespondenceChannel,
    pub status: CorrespondenceChannelStatus,
    pub returned: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContactCorrespondence {
    pub contact_id: Uuid,
    pub identity_id: Uuid,
    pub items: Vec<CorrespondenceItem>,
    pub channels: Vec<CorrespondenceChannelResult>,
    #[serde(default)]
    pub next_cursor: Option<String>,
}

#[cfg(test)]
mod tests {
    use serde_json::{json, Map, Value};

    use super::CorrespondenceItem;

    fn item(channel: &str, fields: Value) -> Value {
        let mut value = Map::from_iter([
            ("channel".into(), channel.into()),
            (
                "source_id".into(),
                "11111111-1111-1111-1111-111111111111".into(),
            ),
            ("direction".into(), "inbound".into()),
            ("occurred_at".into(), "2026-07-20T12:00:00Z".into()),
            (
                "identity_id".into(),
                "22222222-2222-2222-2222-222222222222".into(),
            ),
        ]);
        value.extend(fields.as_object().unwrap().clone());
        Value::Object(value)
    }

    #[test]
    fn parses_each_channel_by_tag() {
        let cases = [
            item(
                "email",
                json!({"mailbox_email": "agent@example.com", "from_address": "person@example.com", "to_addresses": []}),
            ),
            item(
                "sms",
                json!({"conversation_id": "33333333-3333-3333-3333-333333333333", "local_resource_id": "44444444-4444-4444-4444-444444444444", "local_phone_number": "+15550000001", "participants": ["+15550000002"], "matched_contact_phone": "+15550000002", "is_group": false}),
            ),
            item(
                "imessage",
                json!({"conversation_id": "33333333-3333-3333-3333-333333333333", "remote_handle": "+15550000002", "service": "imessage"}),
            ),
            item("calls", json!({"remote_phone_number": "+15550000002"})),
        ];

        assert!(matches!(
            serde_json::from_value(cases[0].clone()).unwrap(),
            CorrespondenceItem::Email(_)
        ));
        assert!(matches!(
            serde_json::from_value(cases[1].clone()).unwrap(),
            CorrespondenceItem::Sms(_)
        ));
        assert!(matches!(
            serde_json::from_value(cases[2].clone()).unwrap(),
            CorrespondenceItem::IMessage(_)
        ));
        assert!(matches!(
            serde_json::from_value(cases[3].clone()).unwrap(),
            CorrespondenceItem::Calls(_)
        ));
    }
}
